### PR TITLE
feat: created crud app

### PR DESCRIPTION
**Pull request by himanshu1243**

### CODE REVIEW

File: Models/studentInfo.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package models
+
+//hello world
```
Comment (rushikesh2) [inline Models/studentInfo.go]: Unnecessary comment

File: Models/studentInfo.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+// Studentinfo represents the model of a student
```
Comment (rushikesh2) [inline Models/studentInfo.go]: It should be StudentInfo.

File: Routes/Routes.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
+package route
+
+//hellow world
```
Comment (rushikesh2) [inline Routes/Routes.go]: Unnecessary comment

File: Routes/Routes.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
+package route
```
Comment (rushikesh2) [inline Routes/Routes.go]: The package name should be routes.

File: Routes/Routes.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+// @Accept  json
+// @Produce  json
+// @Param Studentinfo body StudentInfo true "Create Student"
+// @Success 200 {object} Studentinfo
+// @Router /students [post]
+func AddStudents(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	s := models.StudentInfo{}
```
Comment (rushikesh2) [inline Routes/Routes.go]: Should use appropriate naming conventions for variable.

File: Routes/Routes.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+// @Summary Create a new Student
+// @Description Create a new Student and stores it in the database
+// @Tags Studetn
+// @Accept  json
+// @Produce  json
+// @Param Studentinfo body StudentInfo true "Create Student"
+// @Success 200 {object} Studentinfo
+// @Router /students [post]
+func AddStudents(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
```
Comment (rushikesh2) [inline Routes/Routes.go]: Create Constant package which contains constant.go.
Inside that create constant.go file define constant for "Content-Type", "application/json" use those constants.

File: Routes/Routes.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	s := models.StudentInfo{}
+	err := json.NewDecoder(r.Body).Decode(&s) //error
```
Comment (rushikesh2) [inline Routes/Routes.go]: Unnecessary comment

File: Routes/Routes.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		log.Fatal(err.Error())
+		return
+	}
+	ID, err := strconv.Atoi(s.ID)
+	if err != nil {
+		log.Fatal("conversion not possible")
+		return
+	}
+	_, errin := db.Exec("insert into student(Id,Name,Course) values(?,?,?)", ID, s.Name, s.Course)
```
Comment (rushikesh2) [inline Routes/Routes.go]: instead errin you can use err only

File: Routes/Routes.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	if errin != nil {
+		log.Fatal(err.Error())
+		return
+	}
+	json.NewEncoder(w).Encode(s)
+}
+func DeleteStudent(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
```
Comment (rushikesh2) [inline Routes/Routes.go]: Same as above comment

File: Routes/Routes.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	params := mux.Vars(r)
+	ID, err := strconv.Atoi(params["ID"])
+	if err != nil {
+		log.Fatal(err.Error())
+		return
+	}
+	_, errin := db.Exec("delete from student where ID=?", ID)
```
Comment (rushikesh2) [inline Routes/Routes.go]: same as above comment

File: Routes/Routes.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+// @Router /students [get]
+func GetStudents(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	ss := []models.StudentInfo{}
+	s := models.StudentInfo{}
+	rows, err := db.Query("select*from student")
+	if err != nil {
+		log.Fatal(err.Error()) // log instead of fmt
```
Comment (rushikesh2) [inline Routes/Routes.go]: Unnecessary comment

File: Routes/Routes.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	} else {
```
Comment (rushikesh2) [inline Routes/Routes.go]: Using else you make it code more complex. Remove the else condition and directly perform the logic.

File: Routes/Routes.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+			ss = append(ss, s)
+		}
+		err := json.NewEncoder(w).Encode(ss)
+		if err != nil {
+			log.Fatal(err.Error())
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	}
+	fmt.Print("After error") // error handling
```
Comment (rushikesh2) [inline Routes/Routes.go]: It should be removed.

File: Routes/Routes.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func UpdateStudent(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	s := models.StudentInfo{}
```
Comment (rushikesh2) [inline Routes/Routes.go]: Variable name should be appropriate.

File: SqlConnection/MySqlConnect.go [hunk 15]
```diff
@@ -0,0 +1,18 @@
+package sqlconnection
+
+//hello world
```
Comment (rushikesh2) [inline SqlConnection/MySqlConnect.go]: same as above.

File: SqlConnection/MySqlConnect.go [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"database/sql"
+	"log"
+)
+
+var db *sql.DB
+
+func GetMySQLDB() *sql.DB {
+	db, err := sql.Open("mysql", "root:@(127.0.0.1:3306)/studentinfo")
```
Comment (rushikesh2) [inline SqlConnection/MySqlConnect.go:10]: instead of hardcoding define constant for this.

File: main.go [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+// @version 1.0
+// @description This is a crud application with mysql backend
+
+// @license.name Apache 2.0
+// @license.url http://www.apache.org/licenses/LICENSE-2.0.html
+// @host localhost:8080
+// @BasePath /
+func main() {
+	r := mux.NewRouter()
+	//Read
```
Comment (rushikesh2) [inline main.go]: it should be // Read

File: Routes/Routes.go (hunk 11 above)
Comment (himanshu1243) [reply Routes/Routes.go:83]: which comment?

File: main.go (hunk 17 above)
Comment (himanshu1243) [reply main.go]: didnt understand

File: Routes/Routes.go (hunk 9 above)
Comment (rushikesh2) [reply Routes/Routes.go]: make these string constant then use
"Content-Type", 
"application/json"

File: main.go (hunk 17 above)
Comment (rushikesh2) [reply main.go]: instead of //Read -> // Read

Always keep space after // .

File: Constants/Constants.go [hunk 18]
```diff
@@ -0,0 +1,4 @@
+package constant
```
Comment (rushikesh2) [inline Constants/Constants.go]: The package name is different from than actual package.
actual (constants) -> defined constant

It should be constants.

File: Constants/Constants.go [hunk 19]
```diff
@@ -0,0 +1,4 @@
...
+
+var JsonString = "application/json"
```
Comment (rushikesh2) [inline Constants/Constants.go]: Instead of JsonString you can call as ApplicationJson

File: Constants/Constants.go [hunk 20]
```diff
@@ -0,0 +1,4 @@
...
+var HeaderType = "Content-Type"
```
Comment (rushikesh2) [inline Constants/Constants.go]: Instead of HeaderType you can call as ContentType. It would be more specific.

File: Routes/Routes.go [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+// @Produce  json
+// @Param Studentinfo body StudentInfo true "Create Student"
+// @Success 200 {object} Studentinfo
+// @Router /students [post]
+func AddStudents(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set(constant.HeaderType, constant.JsonString)
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	student := models.StudentInfo{}
+	err := json.NewDecoder(r.Body).Decode(&student) //error
```
Comment (rushikesh2) [inline Routes/Routes.go]: remove unnecessary comment error

File: Routes/Routes.go [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+// @Description Get details of all Students
+// @Tags students
+// @Accept  json
+// @Produce  json
+// @Success 200 {array} Students
+// @Router /students [get]
+func ReadStudents(w http.ResponseWriter, r *http.Request) {
+	db = sqlconnect.GetMySQLDB()
+	defer db.Close()
+	ss := []models.StudentInfo{}
```
Comment (rushikesh2) [inline Routes/Routes.go]: use proper naming convention for variable

File: Routes/Routes.go [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+		ss = append(ss, student)
+	}
+	err = json.NewEncoder(w).Encode(ss)
+	if err != nil {
+		log.Fatal(err.Error())
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	fmt.Print("After error")
```
Comment (rushikesh2) [inline Routes/Routes.go]: not needed.

File: main.go [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+	// Update
+	r.HandleFunc("/student/{ID}", routes.UpdateStudent).Methods("PUT")
+	// Delete
+	r.HandleFunc("/student/{ID}", routes.DeleteStudent).Methods("DELETE")
+	// Swagger
+	r.PathPrefix("/swagger").Handler(httpSwagger.WrapHandler)
+	err := http.ListenAndServe(":8080", r)
+	if err != nil {
+		log.Fatal(err.Error())
+		return
```
Comment (rushikesh2) [inline main.go]: Not needed to return here you can only log the error.

File: Routes/Routes.go [hunk 25]
```diff
@@ -75,7 +75,7 @@ func DeleteStudent(w http.ResponseWriter, r *http.Request) {
 func ReadStudents(w http.ResponseWriter, r *http.Request) {
 	db = sqlconnect.GetMySQLDB()
 	defer db.Close()
-	ss := []models.StudentInfo{}
+	studentslice := []models.StudentInfo{}
```
Comment (rushikesh2) [inline Routes/Routes.go:78]: Make it **studentInfo**. Always keep in mind in go while declaring any variable/function etc, Second word of first letter should be capital.

For example: Studentslice -> StudentSlice,

**Comment by rushikesh2** [comment]
LGTM

**Pull request merged**